### PR TITLE
Add Dotnet V452 install to the PI

### PR DESCRIPTION
**Pull request by gbiellem**
Relates to #179

### CODE REVIEW

File: src/PlatformInstaller/DotNetUpgrade/Events/DotNetDownloadAbortedEvent.cs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+public class DotNetDownloadAbortedEvent
```
Comment (SimonCropp) [inline src/PlatformInstaller/DotNetUpgrade/Events/DotNetDownloadAbortedEvent.cs]: i cant seem to see where this is used

File: src/PlatformInstaller/DotNetUpgrade/Events/DotNetDownloadFailedEvent.cs [hunk 2]
```diff
@@ -0,0 +1,4 @@
+public class DotNetDownloadFailedEvent
+{
+    public string Text { get; set; }
```
Comment (SimonCropp) [inline src/PlatformInstaller/DotNetUpgrade/Events/DotNetDownloadFailedEvent.cs]: the `{ get; set; }` is reundant on all the fields

File: src/PlatformInstaller/DotNetUpgrade/RuntimeUpgradeManager.cs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        this.processRunner = processRunner;
+
+        var tempFolder = new DirectoryInfo(Environment.ExpandEnvironmentVariables("%temp%"));
+        var assetFolder = Directory.CreateDirectory(Path.Combine(tempFolder.FullName, @"Particular\PlatformInstaller"));
+        installer = new FileInfo(Path.Combine(assetFolder.FullName, dotNet452WebInstallerExe));
+    }
+    
+    public bool Is452orLaterInstalled()
+    {
+        using (var ndpKey = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry32).OpenSubKey(@"SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Full\"))
```
Comment (SimonCropp) [inline src/PlatformInstaller/DotNetUpgrade/RuntimeUpgradeManager.cs]: shouldnt the result of ``OpenBaseKey` also be disposed?

File: src/PlatformInstaller/DotNetUpgrade/RuntimeUpgradeManager.cs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                try
+                {
+                    await client.DownloadFileTaskAsync(dotNet452WebInstallerURL, installer.FullName).ConfigureAwait(false);
+                    return installer;
+                }
+                catch
+                {
+                    retries++;
+                    PublishFailed();
+                    Thread.Sleep(500);
```
Comment (SimonCropp) [inline src/PlatformInstaller/DotNetUpgrade/RuntimeUpgradeManager.cs]: i think inside an async  this is better

    await Task.Delay(500).ConfigureAwait(false);

File: src/PlatformInstaller/DotNetUpgrade/RuntimeUpgradeManager.cs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            Text = "Download failed. Retrying..."
+        });
+    }
+
+    void PublishAborted()
+    {
+        eventAggregator.PublishOnUIThread(new DotNetDownloadAbortedEvent
+        {
+            IsError = true,
+            Text = "Download did not complete. Installation Aborted"
```
Comment (SimonCropp) [inline src/PlatformInstaller/DotNetUpgrade/RuntimeUpgradeManager.cs]: isnt the text redundant if we have a specific event for it? or do we intend to re-use this?

File: src/PlatformInstaller/DotNetUpgrade/Views/DotNetDownloadViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
+﻿using Caliburn.Micro;
+
+public  class DotNetDownloadViewModel : Screen, 
+    IHandle<DotNetDownloadProgressEvent>,
+    IHandle<DotNetDownloadStartedEvent>,
+    IHandle<DotNetDownloadCompleteEvent>
+{
+    
+
+    public DotNetDownloadViewModel()
```
Comment (SimonCropp) [inline src/PlatformInstaller/DotNetUpgrade/Views/DotNetDownloadViewModel.cs]: kill the constructor

File: src/PlatformInstaller/DotNetUpgrade/Views/DotNetInstallCompleteViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
+using Caliburn.Micro;
+
+public class DotNetInstallCompleteViewModel : Screen
+{
+
+    public DotNetInstallCompleteViewModel()
```
Comment (SimonCropp) [inline src/PlatformInstaller/DotNetUpgrade/Views/DotNetInstallCompleteViewModel.cs]: kill the constructor

File: src/PlatformInstaller/NativeMethods/NativeConstants.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
+﻿internal static class NativeConstants
```
Comment (SimonCropp) [inline src/PlatformInstaller/NativeMethods/NativeConstants.cs]: kill the internal

File: src/PlatformInstaller/NativeMethods/ShowWindowCommands.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
+﻿internal enum ShowWindowCommands : int
```
Comment (SimonCropp) [inline src/PlatformInstaller/NativeMethods/ShowWindowCommands.cs]: assume this was copied from somewhere? should we link to that?

File: src/PlatformInstaller/Views/ShellViewModel.cs [hunk 10]
```diff
@@ -119,6 +125,12 @@ void RunStartupSequence()
                 }
             }
         }
+
+        if (!runtimeUpgradeManager.Is452orLaterInstalled())
```
Comment (SimonCropp) [inline src/PlatformInstaller/Views/ShellViewModel.cs]: perhaps invert this if? makes for easier reading if u dont need to notice the negation

File: src/PlatformInstaller/Extensions/TaskExtensions.cs [hunk 11]
```diff
@@ -1,35 +0,0 @@
-﻿using System;
```
Comment (SimonCropp) [inline src/PlatformInstaller/Extensions/TaskExtensions.cs]: i think we can also delete the dir from the project?

File: src/PlatformInstaller/DotNetUpgrade/Views/DotNetPreReqView.xaml [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+        </ResourceDictionary>
+    </UserControl.Resources>
+    <DockPanel >
+        <TextBlock DockPanel.Dock="Top" FontSize="18" FontWeight="Bold">Install .NET Framework 4.5.2</TextBlock>
+        <StackPanel DockPanel.Dock="Bottom" Orientation="Horizontal" FlowDirection="RightToLeft">
+            <Button x:Name="Exit" Content="Exit"/>
+            <Button x:Name="Install" Content="Install"/>
+        </StackPanel>
+        <StackPanel>
+            <TextBlock Foreground="Gray">Please click Install to commence the download and installation of this prerequisite. </TextBlock>
```
Comment (SimonCropp) [inline src/PlatformInstaller/DotNetUpgrade/Views/DotNetPreReqView.xaml]: perhaps give them a link here for info of how to install manually?

File: src/PlatformInstaller/DotNetUpgrade/Views/DotNetInstallFailedView.xaml [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+                <ResourceDictionary Source="../../Styles.xaml"/>
+            </ResourceDictionary.MergedDictionaries>
+        </ResourceDictionary>
+    </UserControl.Resources>
+    <DockPanel >
+        <TextBlock DockPanel.Dock="Top" FontSize="18" FontWeight="Bold">Failed to Download or Install .NET Framework 4.5.2</TextBlock>
+        <StackPanel DockPanel.Dock="Bottom" Orientation="Horizontal" FlowDirection="RightToLeft">
+            <Button x:Name="Exit" Content="Exit"/>
+        </StackPanel>
+        <TextBlock Foreground="Gray">Please install .NET Framework 4.5.2 manually. The installation can be found on the Microsoft Web Site <global:Link NavigateUri="https://www.microsoft.com/en-us/download/details.aspx?id=42642">here</global:Link></TextBlock>
```
Comment (SimonCropp) [inline src/PlatformInstaller/DotNetUpgrade/Views/DotNetInstallFailedView.xaml]: perhaps 
```
<TextBlock Foreground="Gray">Please install .NET Framework 4.5.2 manually. The installation can be found on the <global:Link NavigateUri="https://www.microsoft.com/en-us/download/details.aspx?id=42642">Microsoft Download Site</global:Link>.</TextBlock>
```

File: src/PlatformInstaller/DotNetUpgrade/RuntimeUpgradeManager.cs [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+public class RuntimeUpgradeManager
+{
+    IEventAggregator eventAggregator;
+    CredentialStore credentialStore;
+    ProcessRunner processRunner;
+    const string dotNet452WebInstallerExe = "NDP452-KB2901954-Web.exe";
+
+    // This download link came from .NET Framework Deployment Guide for Developers -  see https://msdn.microsoft.com/en-us/library/ee942965%28v=vs.110%29.aspx#redist
+    const string dotNet452WebInstallerURL = @"http://go.microsoft.com/fwlink/?LinkId=397707";
```
Comment (SimonCropp) [inline src/PlatformInstaller/DotNetUpgrade/RuntimeUpgradeManager.cs]: minor but i think the @ is redundant

File: src/PlatformInstaller/DotNetUpgrade/RuntimeUpgradeManager.cs (hunk 5 above)
Comment (gbiellem) [inline src/PlatformInstaller/DotNetUpgrade/RuntimeUpgradeManager.cs]: :+1:

File: src/PlatformInstaller/NativeMethods/ShowWindowCommands.cs (hunk 9 above)
Comment (gbiellem) [inline src/PlatformInstaller/NativeMethods/ShowWindowCommands.cs]: I copied it from SC installer,  There was no link to the origin in that code base

File: src/PlatformInstaller/Extensions/TaskExtensions.cs (hunk 11 above)
Comment (gbiellem) [inline src/PlatformInstaller/Extensions/TaskExtensions.cs]: done

File: src/PlatformInstaller/DotNetUpgrade/Events/DotNetDownloadAbortedEvent.cs (hunk 1 above)
Comment (gbiellem) [inline src/PlatformInstaller/DotNetUpgrade/Events/DotNetDownloadAbortedEvent.cs]: WIP - No wired up yet.

Comment (gbiellem) [inline src/PlatformInstaller/DotNetUpgrade/Events/DotNetDownloadAbortedEvent.cs]: It's wired up now

File: src/PlatformInstaller/DotNetUpgrade/Views/DotNetPreReqView.xaml (hunk 12 above)
Comment (gbiellem) [inline src/PlatformInstaller/DotNetUpgrade/Views/DotNetPreReqView.xaml]: Not sure.   We provide the link on failure to download or install. I thought that would be enough

**Comment by gbiellem** [comment]
@SimonCropp  Please take another look.

File: src/PlatformInstaller/DotNetUpgrade/Views/DotNetPreReqView.xaml (hunk 12 above)
Comment (SimonCropp) [inline src/PlatformInstaller/DotNetUpgrade/Views/DotNetPreReqView.xaml]: i am ok with that

**Comment by SimonCropp** [comment]
looks good to me. want me to do some smoke testing tomorrow?

**Comment by gbiellem** [comment]
@SimonCropp Sure. I'll be doing the same.

**Comment by gbiellem** [comment]
@SimonCropp  noticed I got an unhandled exception on download failure during the smoke testing .  I've fixed it.  Seems you hit this already in #177

**Pull request merged**